tests/scenarios/registration.spec.ts

```diff
@@ -0,0 +1,455 @@
+      termsOfServiceHash: ProofOfWork.hashAsHexString([termsOfService]),
+    };
+
+    const responseNonce = ProofOfWork.findQualifiedResponseNonce({
+      challengeNonce,
+      maximumAllowedHashValue,
+      requestData: JSON.stringify(registrationData),
+    });
+
+    // 3. Alice sends the registration request to the server and it is rejected.
+    const registrationRequest: RegistrationRequest = {
+      registrationData,
+      proofOfWork: {
+        challengeNonce,
+        responseNonce,
+      },
+    };
+    
+    const registrationResponse = await fetch(registrationEndpoint, {
+      method: 'POST',
+      headers: { 'Content-Type': 'application/json' },
+      body: JSON.stringify(registrationRequest),
+    });
+    const registrationResponseBody = await registrationResponse.json() as any;
+    expect(registrationResponse.status).to.equal(400);
+    expect(registrationResponseBody.code).to.equal(DwnServerErrorCode.ProofOfWorkManagerInvalidChallengeNonce);
+  });
+
+  it('should reject a DWN message for an existing tenant who agreed to an outdated terms-of-service.', async () => {
+    // Scenario:
+    // 1. Alice is a registered tenant and is able to write to the DWN.
+    // 2. DWN server administrator updates the terms-of-service.
+    // 3. Alice no longer can write to the DWN because she has not agreed to the new terms-of-service.
+
+    // 1. Alice is a registered tenant and is able to write to the DWN.
+    // Short-cut to register Alice.
+    registrationManager.recordTenantRegistration({
+      did: alice.did,
+      termsOfServiceHash: ProofOfWork.hashAsHexString([registrationManager.getTermsOfService()])
+    });
+
+    // Sanity test that Alice can write to the DWN after registration.
+    const write1 = await generateRecordsWriteJsonRpcRequest(alice);
+    const write1Response = await fetch(dwnMessageEndpoint, {
+      method: 'POST',
+      headers: {
+        'dwn-request': JSON.stringify(write1.jsonRpcRequest),
+      },
+      body: new Blob([write1.dataBytes]),
+    });
+    const write1ResponseBody = await write1Response.json() as JsonRpcResponse;
+    expect(write1Response.status).to.equal(200);
+    expect(write1ResponseBody.result.reply.status.code).to.equal(202);
+
+    // 2. DWN server administrator updates the terms-of-service.
+    registrationManager.updateTermsOfService('new terms of service');
+
+    // 3. Alice no longer can write to the DWN because she has not agreed to the new terms-of-service.
+    const write2 = await generateRecordsWriteJsonRpcRequest(alice);
+    const write2Response = await fetch(dwnMessageEndpoint, {
+      method: 'POST',
+      headers: {
+        'dwn-request': JSON.stringify(write2.jsonRpcRequest),
+      },
+      body: new Blob([write2.dataBytes]),
+    });
+    const write2ResponseBody = await write2Response.json() as JsonRpcResponse;
+    expect(write2Response.status).to.equal(200);
+    expect(write2ResponseBody.result.reply.status.code).to.equal(401);
```

Comment:
Sanity: consider adding to this test where Alice re-registers and can now write the same DWN message.

Reply:
Good call! Adding now.